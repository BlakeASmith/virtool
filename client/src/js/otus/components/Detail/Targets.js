import { find, get, map } from "lodash-es";
import React from "react";
import { connect } from "react-redux";
import { BoxGroup } from "../../../base";
import { checkRefRight, formatIsolateName } from "../../../utils/utils";
import { showAddSequence, showEditSequence, showRemoveSequence } from "../../actions";
import { getActiveIsolate, getSequences } from "../../selectors";
import AddSequence from "./Sequences/Add";
import EditSequence from "./Sequences/Edit";
import RemoveSequence from "./Sequences/Remove";
import Sequence from "./Sequences/Sequence";
import { Target } from "./Target";

export const IsolateTargets = props => {
    const required = props.target === true ? "REQUIRED" : "NOT REQUIRED";

    let targetComponents;

    if (props.targets) {
        targetComponents = map(props.targets, target => {
            const sequence = find(props.sequences, { target: target.name });

            if (sequence != null) {
                return (
                    <Sequence
                        key={sequence.id}
                        active={sequence.accession === props.activeSequenceId}
                        canModify={props.canModify}
                        showEditSequence={props.showEditSequence}
                        showRemoveSequence={props.showRemoveSequence}
                        targetName={sequence.target}
                        description={target.description}
                        length={target.length}
                        required={required}
                        name={target.name}
                        {...sequence}
                    />
                );
            }

            return (
                <Target
                    key={target.name}
                    name={target.name}
                    required={required}
                    description={target.description}
                    length={target.length}
                    onClick={() => props.showAddSequence(target.name)}
                />
            );
        });
    }

    const targetDetail = find(props.targets, { name: props.targetName });
    const addSequence = <AddSequence {...targetDetail} required={required} />;

    return (
        <React.Fragment>
            <label>Targets</label>
            <BoxGroup>{targetComponents}</BoxGroup>

            <EditSequence otuId={props.otuId} isolateId={props.activeIsolateId} error={props.error} />
            <RemoveSequence otuId={props.otuId} isolateId={props.activeIsolateId} isolateName={props.isolateName} />
            {addSequence}
        </React.Fragment>
    );
};

const mapStateToProps = state => {
    const activeIsolateId = state.otus.activeIsolateId;

    const activeIsolate = getActiveIsolate(state);
    const sequences = getSequences(state);

    return {
        activeIsolateId,
        sequences,
        targetName: state.otus.targetName,
        otuId: state.otus.detail.id,
        editing: state.otus.editSequence,
        isolateName: formatIsolateName(activeIsolate),
        canModify: !get(state, "references.detail.remotes_from") && checkRefRight(state, "modify_otu"),
        error: get(state, "errors.EDIT_SEQUENCE_ERROR.message", ""),
        targets: state.references.detail.targets
    };
};

const mapDispatchToProps = dispatch => ({
    showAddSequence: targetName => {
        dispatch(showAddSequence(targetName));
    },

    showEditSequence: sequenceId => {
        dispatch(showEditSequence(sequenceId));
    },

    showRemoveSequence: sequenceId => {
        dispatch(showRemoveSequence(sequenceId));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(IsolateTargets);
