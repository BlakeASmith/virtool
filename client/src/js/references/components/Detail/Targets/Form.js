import React from "react";
import styled from "styled-components";
import { Input } from "../../../../base";

const StyledRequire = styled.div`
    display: flex;
    flex-direction: row;
    align-items: center;
`;
export const TargetForm = ({ onChange, name, description, length, required }) => {
    return (
        <div>
            <Input label="Name" name="name" value={name} onChange={onChange} />
            <Input label="Description" name="description" value={description} onChange={onChange} />
            <Input type="number" label="Length" name="length" value={length} onChange={onChange} />
            <StyledRequire>
                <input type="checkbox" name="required" checked={required} onChange={onChange} />
                Required
            </StyledRequire>
        </div>
    );
};
