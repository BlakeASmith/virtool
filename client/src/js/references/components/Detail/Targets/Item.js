import React from "react";
import styled from "styled-components";
import { BoxGroupSection, Icon, Flex, FlexItem } from "../../../../base";

const NameIcon = styled.div`
    display: flex;
    justify-content: space-between;
`;

export const TargetItem = ({ name, description, onEdit, onRemove }) => {
    return (
        <BoxGroupSection>
            <NameIcon>
                {name}
                <FlexItem grow={1} shrink={1}>
                    <Flex alignItems="center" className="pull-right">
                        <Icon name="edit" bsStyle="warning" tip="Modify" onClick={onEdit} />
                        <FlexItem pad>
                            <Icon name="trash" bsStyle="danger" tip="Remove" onClick={onRemove} />
                        </FlexItem>
                    </Flex>
                </FlexItem>
            </NameIcon>
            {description}
        </BoxGroupSection>
    );
};
