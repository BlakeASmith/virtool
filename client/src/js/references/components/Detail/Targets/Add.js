import React from "react";
import { Modal } from "react-bootstrap";
import { connect } from "react-redux";
import { Button } from "../../../../base";
import { editReference } from "../../../actions";
import { TargetForm } from "./Form";

const getInitialState = () => ({
    name: "",
    description: "",
    length: 0,
    required: false,
    errorName: ""
});

export class AddTarget extends React.Component {
    constructor(props) {
        super(props);
        this.state = getInitialState();
    }

    handleSubmit = e => {
        e.preventDefault();

        if (!this.state.name) {
            this.setState({ errorName: "Required Field" });
        }

        const newTarget = [
            ...this.props.targets,
            {
                name: this.state.name,
                description: this.state.description,
                length: new Number(this.state.length),
                required: this.state.required
            }
        ];

        const update = {
            targets: newTarget
        };

        if (this.state.name) {
            this.props.onSubmit(this.props.refId, update);
        }

        this.props.onHide();
    };

    handleChange = e => {
        this.setState({
            [e.target.name]: e.target.value
        });
    };

    handleClick = () => {
        this.setState({
            required: !this.state.required
        });
    };

    handleExited = () => {
        this.props.onHide();
        this.setState(getInitialState());
    };

    render() {
        return (
            <Modal
                show={this.props.show}
                onHide={this.props.onHide}
                onEnter={this.handleEnter}
                onExited={this.handleExited}
            >
                <Modal.Header closeButton>
                    <span className="text-capitalize">Add target</span>
                </Modal.Header>
                <form onSubmit={this.handleSubmit}>
                    <Modal.Body>
                        <TargetForm
                            onChange={this.handleChange}
                            name={this.state.name}
                            description={this.state.description}
                            length={this.state.length}
                            required={this.state.required}
                            errorName={this.errorName}
                            onClick={this.handleClick}
                        />
                    </Modal.Body>
                    <Modal.Footer>
                        <Button type="submit" icon="save" bsStyle="primary">
                            Submit
                        </Button>
                    </Modal.Footer>
                </form>
            </Modal>
        );
    }
}

export const mapStateToProps = state => ({
    names: state.references.detail.name,
    dataType: state.references.detail.data_type,
    documents: state.references.documents,
    refId: state.references.detail.id,
    targets: state.references.detail.targets
});

export const mapDispatchToProps = dispatch => ({
    onSubmit: (refId, update) => {
        dispatch(editReference(refId, update));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(AddTarget);
