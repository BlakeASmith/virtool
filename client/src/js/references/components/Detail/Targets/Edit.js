import React from "react";

import { connect } from "react-redux";
import { findIndex, find } from "lodash-es";
import { Button, ModalDialog, DialogBody, DialogFooter } from "../../../../base";
import { editReference } from "../../../actions";
import { TargetForm } from "./Form";

const getInitialState = ({ initialName, initialDescription, initialLength, initialRequired }) => ({
    name: initialName || "",
    description: initialDescription || "",
    length: initialLength || 0,
    required: initialRequired || false,
    errorName: ""
});

export class EditTarget extends React.Component {
    constructor(props) {
        super(props);
        this.state = getInitialState(this.props);
    }

    handleSubmit = e => {
        e.preventDefault();

        const targets = this.props.targets;
        const initialTarget = this.props.initialTarget;
        const newTarget = [...this.props.targets];
        newTarget.splice(findIndex(targets, initialTarget), 1, {
            name: this.state.name,
            description: this.state.description,
            length: new Number(this.state.length),
            required: this.state.required
        });

        const update = {
            targets: newTarget
        };

        if (this.state.name) {
            this.props.onSubmit(this.props.refId, update);
        }
        this.props.onHide();
    };

    handleChange = e => {
        this.setState({
            [e.target.name]: e.target.value,
            required: e.target.checked
        });
    };

    handleEnter = () => {
        this.setState(getInitialState(this.props));
    };

    render() {
        return (
            <ModalDialog
                headerText="Edit target"
                text-capitalize="text-capitalize"
                show={this.props.show}
                onHide={this.props.onHide}
                onEnter={this.handleEnter}
                onExited={this.handleExited}
            >
                <DialogBody>
                    <form onSubmit={this.handleSubmit}>
                        <TargetForm
                            onChange={this.handleChange}
                            name={this.state.name}
                            description={this.state.description}
                            length={this.state.length}
                            required={this.state.required}
                            errorName={this.errorName}
                        />
                        <DialogFooter>
                            <Button type="submit" icon="save" bsStyle="primary">
                                Submit
                            </Button>
                        </DialogFooter>
                    </form>
                </DialogBody>
            </ModalDialog>
        );
    }
}

export const mapStateToProps = (state, ownProps) => {
    const activeName = ownProps.activeName;

    let initialTarget = {};

    if (activeName) {
        initialTarget = find(state.references.detail.targets, ["name", ownProps.activeName]);
        if (!initialTarget) {
            initialTarget = {};
        }
    }

    const { name, description, length, required } = initialTarget;

    return {
        initialTarget,
        targets: state.references.detail.targets,
        initialName: name,
        initialDescription: description,
        initialLength: length,
        initialRequired: required,
        refId: state.references.detail.id
    };
};

export const mapDispatchToProps = dispatch => ({
    onSubmit: (refId, update) => {
        dispatch(editReference(refId, update));
    }
});

export default connect(mapStateToProps, mapDispatchToProps)(EditTarget);
