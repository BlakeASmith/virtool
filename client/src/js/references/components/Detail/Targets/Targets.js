import React from "react";
import styled from "styled-components";
import { connect } from "react-redux";
import { map, filter } from "lodash-es";
import { BoxGroupHeader, BoxGroup } from "../../../../base";
import { editReference } from "../../../actions";
import AddTarget from "./Add";
import EditTarget from "./Edit";
import { TargetItem } from "./Item";

const StyledAddTargets = styled(BoxGroupHeader)`
    display: flex;
    flex-direction: row;
    justify-content: space-between;
`;

const getInitialState = () => ({
    showAdd: false,
    showEdit: false
});

const StyledAddTargetsButton = styled.a`
    cursor: pointer;
`;

export class Targets extends React.Component {
    constructor(props) {
        super(props);
        this.state = getInitialState();
    }

    handleHide = () => {
        this.setState({ showAdd: false, showEdit: false });
    };

    add = () => {
        this.setState({ showAdd: true });
    };

    edit = name => {
        this.setState({ showEdit: true, activeName: name });
    };

    handleRemove = name => {
        const update = {
            targets: filter([...this.props.targets], { name })
        };

        this.props.onRemove(this.props.refId, update);
    };

    render() {
        const addButton = <StyledAddTargetsButton onClick={this.add}>Add target</StyledAddTargetsButton>;

        const targetComponents = map(this.props.targets, target => (
            <TargetItem
                key={target.name}
                {...target}
                onEdit={() => this.edit(target.name)}
                onRemove={() => this.handleRemove(target.name)}
            />
        ));

        return (
            <div>
                <StyledAddTargets>
                    Targets
                    {addButton}
                </StyledAddTargets>
                <BoxGroup>{targetComponents}</BoxGroup>
                <AddTarget show={this.state.showAdd} onHide={this.handleHide} />
                <EditTarget show={this.state.showEdit} onHide={this.handleHide} activeName={this.state.activeName} />
            </div>
        );
    }
}

const mapStateToProps = state => ({
    ...state.references.detail.targets,
    refId: state.references.detail.id,
    targets: state.references.detail.targets
});

const mapDispatchToProps = dispatch => ({
    onRemove: (refId, update) => {
        dispatch(editReference(refId, update));
    }
});
export default connect(mapStateToProps, mapDispatchToProps)(Targets);
