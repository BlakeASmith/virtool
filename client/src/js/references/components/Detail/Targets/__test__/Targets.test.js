import { mapStateToProps, mapDispatchToProps, Targets, StyledAddTargetsButton } from "../Targets";

describe("<Targets />", () => {
    const props = {
        targets: [{ name: "foo" }],
        onRemove: jest.fn(),
        refId: "bar"
    };

    it("should render", () => {
        const wrapper = shallow(<Targets {...props} />);
        expect(wrapper).toMatchSnapshot();
    });

    it("add() should update state", () => {
        const wrapper = shallow(<Targets {...props} />);
        wrapper.find(StyledAddTargetsButton).simulate("click");
        expect(wrapper.state()).toEqual({ showAdd: true, showEdit: false });
    });

    it("edit() should update state", () => {
        const wrapper = shallow(<Targets {...props} />);
        wrapper.find("TargetItem").simulate("edit");
        expect(wrapper.state()).toEqual({
            activeName: "foo",
            showAdd: false,
            showEdit: true
        });
    });

    it("handleRemove() should call onRemove()", () => {
        const wrapper = shallow(<Targets {...props} />);
        wrapper.find("TargetItem").simulate("remove");
        expect(props.onRemove).toHaveBeenCalledWith("bar", { targets: [] });
    });
});

describe("mapStateToProps()", () => {
    const state = {
        references: {
            detail: {
                id: "baz",
                targets: [
                    { name: "foo", description: "bar", required: false },
                    { name: "Foo", description: "Bar", required: true }
                ]
            }
        }
    };
    it("should return props", () => {
        const result = mapStateToProps(state);
        expect(result).toEqual({
            0: { name: "foo", description: "bar", required: false },
            1: { name: "Foo", description: "Bar", required: true },

            refId: "baz",
            targets: [
                { name: "foo", description: "bar", required: false },
                { name: "Foo", description: "Bar", required: true }
            ]
        });
    });
});

describe("mapDispatchToProps()", () => {
    it("should return onRemove in props ", () => {
        const dispatch = jest.fn();
        const props = mapDispatchToProps(dispatch);
        props.onRemove("foo", "bar");
        expect(dispatch).toHaveBeenCalledWith({ refId: "foo", update: "bar", type: "EDIT_REFERENCE_REQUESTED" });
    });
});
