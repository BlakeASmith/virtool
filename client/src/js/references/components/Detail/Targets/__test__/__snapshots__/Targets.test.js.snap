// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Targets /> should render 1`] = `
<Box__BoxGroup>
  <Targets__StyledAddTargets>
    <h2>
      Targets
    </h2>
    <Targets__StyledAddTargetsButton
      onClick={[Function]}
    >
      Add target
    </Targets__StyledAddTargetsButton>
  </Targets__StyledAddTargets>
  <div>
    <TargetItem
      key="foo"
      name="foo"
      onEdit={[Function]}
      onRemove={[Function]}
    />
  </div>
  <Connect(AddTarget)
    onHide={[Function]}
    show={false}
  />
  <Connect(EditTarget)
    onHide={[Function]}
    show={false}
  />
</Box__BoxGroup>
`;
